proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,424 @@
+# Extend Property Wrappers to Function and Closure Parameters
+
+* Proposal: [SE-NNNN](NNNN-extend-property-wrappers-to-function-and-closure-parameters.md)
+* Authors: [Holly Borla](https://github.com/hborla), [Filip Sakel](https://github.com/filip-sakel)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Implementation: [apple/swift#34272](https://github.com/apple/swift/pull/34272)
+
+
+## Introduction
+
+Property Wrappers were [introduced in Swift 5.1](https://github.com/apple/swift-evolution/blob/main/proposals/0258-property-wrappers.md), and have since become a popular feature abstracting away common accessor patterns for properties. Currently, applying a property wrapper is solely permitted at local and type context. However, with increasing adoption, demand for extending _where_ property wrappers can be applied has emerged. This proposal aims to extend property wrappers to function and closure parameters.
+
+
+## Motivation
+
+Property wrappers have undoubtably been very successful. Applying a property wrapper to a property is enabled by an incredibly lightweight and expressive syntax. Therefore, library authors can expose complex behavior through easily understandable property-wrapper types in an efficient manner. For instance, frameworks such as [SwiftUI](https://developer.apple.com/documentation/swiftui/) and [Combine](https://developer.apple.com/documentation/combine) introduce property wrappers such as [`State`](https://developer.apple.com/documentation/swiftui/state), [`Binding`](https://developer.apple.com/documentation/swiftui/binding) and [`Published`](https://developer.apple.com/documentation/combine/published) respectively, to expose elaborate behavior through a succinct interface, helping craft expressive yet simple APIs. However, property wrappers are only applicable to local variables and type properties, shattering the illusion that they helped realize in the first place when working with parameters.
+
+### Memberwise initialization
+
+Currently, property-wrapped struct properties interact with function parameters, through the struct's synthesized memberwise initializer. However, property-wrapper attributes are _not_ supported on function parameters. This leads to really complicated and nuanced rules for which type, between the wrapped type and the backing property-wrapper type, the memberwise initializer accepts. Namely, the compiler will choose the wrapped-value type to offer a convenience to the call-site, when the property wrapper supports initalization through the suitable initializer: `init(wrappedValue:)`, as seen here:
```

Comment:
> "complicated and nuanced rules for which type the memberwise initializer accepts"

After trying to wrap my head around what you mean by this, I think I figured it out. So, I changed this paragraph in order to simplify it as much as I could. Let me know if you think I missed something; also, any changes that simplify this portion are definitely welcome (especially because this is the first paragraph of Motivation).

Reply:
Clarifying that the compiler is choosing between the wrapped type and the backing wrapper type is great, thank you! 

We're trying to eliminate the phrase "suitable initializer" or "suitable `init(wrappedValue:)`", though, so I'll replace that part (sorry if it was me who originally wrote that 🙂 ).

Reply:
It's okey, don't worry about that. I think it is much better know, anyway!